base/strings/string.jl

```diff
@@ -145,6 +145,111 @@ end
+    # which is the second lowest state shift.
+    shifts = [0, 13, 6, 10, 4]
+
+    # Both of these state tables are only 4 states wide even though there are 5 states
+    # because the machine must be reset once it is in state 4
+    forward_state_table  = [    [0,  4,  4,  4],
+                                [4,  0,  1,  2],
+                                [1,  4,  4,  4],
+                                [2,  4,  4,  4],
+                                [3,  4,  4,  4],
+                                [4,  4,  4,  4] ]
+
+    reverse_state_table  = [    [0,  4,  4,  4],
+                                [1,  2,  3,  4],
+                                [0,  0,  4,  4],
+                                [0,  0,  0,  4],
+                                [0,  0,  0,  0],
+                                [4,  4,  4,  4] ]
+
+
+    f(from, to) = _GUTF8State(shifts[to + 1]) << shifts[from + 1]
+    r(state_row) = |([f(n - 1, state_row[n]) for n in 1:length(state_row)]...)
+    forward_class_rows = [r(forward_state_table[n]) for n in 1:length(forward_state_table)]
+    reverse_class_rows = [r(reverse_state_table[n]) for n in 1:length(reverse_state_table)]
+
+    byte_class = [  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    #  0x00:0x0F      00000000:00001111
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    #  0x10:0x1F      00010000:00011111
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    #  0x20:0x2F      00100000:00101111
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    #  0x30:0x3F      00110000:00111111
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    #  0x40:0x4F      01000000:01001111
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    #  0x50:0x5F      01010000:01011111
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    #  0x60:0x6F      01100000:01101111
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    #  0x70:0x7F      01110000:01111111
+                    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,    #  0x80:0x8F      10000000:10001111
+                    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,    #  0x90:0x9F      10010000:10011111
+                    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,    #  0xA0:0xAF      10100000:10101111
+                    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,    #  0xB0:0xBF      10110000:10111111
+                    2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,    #  0xC0:0xCF      11000000:11001111
+                    2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,    #  0xD0:0xDF      11010000:11011111
+                    3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,    #  0xE0:0xEF      11100000:11101111
+                    4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5  ]  #  0xF0:0xFF      11110000:11111111
+    forward_dfa_table = zeros(_GUTF8State, 256)
```

Comment:
It might be better to have these be tuples.

Reply:
Are tuples with 256 element performant?

Reply:
at the very least, this should probably be a `Memory` rather than a `Vector` (it should be slightly faster)

Reply:
I don't see why they wouldn't be in this case when they just act as a global.

Reply:
With it being a tuple there's a chance some of it gets folded away, while with the array it won't.